File: frontend/src/types/oc.ts
export type EstadoOC =
  | "nueva"
  | "en_cotizacion"
  | "pendiente_aprobacion"
  | "aprobada"
  | "rechazada"
  | "cancelada"
  | "en_correccion"
  | "oc_enviada"
  | "oc_en_plataforma"
  | "entregada"
  | "cerrada"

export interface SolicitudOC {
  id: string
  consecutivo_os: string
  descripcion: string
  categoria: string | null
  grupo_articulos: string | null
  cantidad: number
  nivel_prioridad: string
  solicitante_nombre: string
  solicitante_email: string | null
  area_solicitante: string | null
  sede: string | null
  cliente: string | null
  condicion: string | null
  observaciones_solicitante: string | null
  placa_ficha: string | null
  tipo_solicitud: "compra" | "mantenimiento"
  tipo_mantenimiento: string | null
  clasificacion_mantenimiento: string | null
  fecha_proximo_mantenimiento: string | null
  origen_solicitud_id: string | null
  archivada: boolean
  tiene_proforma: boolean
  proforma_path: string | null
  estado: EstadoOC
  auxiliar_id: number | null
  evidencia_url: string | null
  plataforma: string | null
  numero_remision: string | null
  observaciones_compras: string | null
  fecha_estimada_entrega: string | null
  fecha_confirmada_entrega: string | null
  numero_factura: string | null
  aval_compra: string | null
  observacion_contabilidad: string | null
  fecha_recibida_factura: string | null
  fotos_producto: string[] | null
  fecha_solicitud: string
  fecha_asignacion: string | null
  fecha_cotizacion: string | null
  fecha_aprobacion: string | null
  fecha_envio_oc: string | null
  fecha_en_plataforma: string | null
  fecha_recibido: string | null
  fecha_cerrado: string | null
  created_at: string
  updated_at: string
}

/** Respuesta de GET /api/oc/solicitudes (listado paginado). */
export interface SolicitudesListResponse {
  items: SolicitudOC[]
  total: number
}

export interface ItemCotizacion {
  num?: number
  descripcion: string
  referencia?: string
  cantidad?: number
  valor_unitario?: number
  valor_total?: number
}

export interface CotizacionProveedor {
  id: string
  solicitud_id: string
  proveedor_nombre: string
  proveedor_nit: string | null
  proveedor_email: string | null
  numero_cotizacion_proveedor: string | null
  valor_unitario: number
  valor_antes_iva: number | null
  valor_iva: number | null
  valor_total: number
  fecha_estimada_entrega: string | null
  forma_pago: string | null
  plazo_entrega: string | null
  garantia: string | null
  anticipo: string | null
  pago_saldo: string | null
  observaciones: string | null
  pdf_path: string | null
  extraccion_automatica: boolean
  aprobada: boolean | null
  valor_aprobado: number | null
  valor_aprobado_original: number | null
  aprobado_por_id: number | null
  observaciones_aprobacion: string | null
  items: ItemCotizacion[] | null
  created_at: string
}

export interface HistorialEntrada {
  id: string
  estado_anterior: string | null
  estado_nuevo: string
  usuario_nombre: string | null
  notas: string | null
  fecha: string
  es_reproceso?: boolean
  tipo_accion?: string | null
}

export interface OrdenCompra {
  id: string
  solicitud_id: string
  cotizacion_id: string
  numero_oc: string
  pdf_path: string | null
  enviada_proveedor: boolean
  enviada_coordinador: boolean
  email_proveedor: string | null
  created_at: string
}

export interface Proveedor {
  id: string
  nombre: string
  email: string | null
  telefono: string | null
  nit: string | null
  categoria: string | null
  activo: boolean
  created_at: string
}

export interface ConteoItem {
  label: string
  count: number
}

export interface MesItem {
  mes: string           // "2026-04"
  label: string         // "Abr 2026"
  solicitudes: number
  valor_aprobado: number
}

export interface UsuarioBasico {
  id: number
  full_name: string
  email: string
  area: string | null
  role: string
}

export interface KPIData {
  total_solicitudes: number
  por_estado: ConteoItem[]
  por_plataforma: ConteoItem[]
  por_prioridad: ConteoItem[]
  por_area: ConteoItem[]
  valor_total_aprobado: number
  valor_total_sin_iva: number
  valor_iva_acumulado: number
  total_ordenes_generadas: number
  top_proveedores: ConteoItem[]
  /** Promedio días desde fecha_solicitud hasta fecha_asignacion (solo solicitudes ya asignadas en el filtro). */
  tiempo_promedio_asignacion_dias: number
  muestras_asignacion: number
  /** Promedio días desde entrada a en_correccion hasta la siguiente transición (solo ciclos cerrados). */
  tiempo_promedio_correccion_solicitante_dias: number
  ciclos_correccion_resueltos: number
  tiempo_promedio_cotizacion_dias: number
  por_mes: MesItem[]
  solicitudes_recientes: {
    id: string
    consecutivo_os: string
    descripcion: string
    estado: string
    nivel_prioridad: string
    plataforma: string | null
    fecha_solicitud: string
  }[]
  reprocesos_total: number
  tiempo_promedio_reproceso_dias: number
  correcciones_directivo: number
  rechazos_solicitud: number
  rechazos_cotizacion: number
  /** Presente en backends recientes; si falta, el frontend puede sintetizarlo con `resolverReporteTiemposKpis`. */
  reporte_tiempos?: {
    texto_para_informe: string
    metricas: {
      clave: string
      etiqueta: string
      subtitulo: string
      valor: number
      unidad: string
      ayuda: string
    }[]
    generado_en_utc: string
    nota_metodologia: string
    sugerencia_agentes: string
  }
}

File: backend/app/models/oc.py
import uuid
from datetime import date, datetime, timezone
from enum import Enum
from typing import Optional

from sqlmodel import Field, SQLModel


class EstadoOC(str, Enum):
    nueva = "nueva"
    en_cotizacion = "en_cotizacion"
    pendiente_aprobacion = "pendiente_aprobacion"
    aprobada = "aprobada"
    rechazada = "rechazada"
    oc_enviada = "oc_enviada"
    oc_en_plataforma = "oc_en_plataforma"
    entregada = "entregada"
    cerrada = "cerrada"


class SolicitudOC(SQLModel, table=True):
    __tablename__ = "oc_solicitudes"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)

    # Datos que vienen de SharePoint/Forms via webhook de PA
    consecutivo_os: str = Field(max_length=50)
    descripcion: str
    categoria: Optional[str] = Field(default=None, max_length=100)
    grupo_articulos: Optional[str] = Field(default=None, max_length=100)
    cantidad: int
    nivel_prioridad: str = Field(default="Media", max_length=20)
    solicitante_nombre: str = Field(max_length=200)
    solicitante_email: Optional[str] = Field(default=None, max_length=200)
    area_solicitante: Optional[str] = Field(default=None, max_length=100)
    sede: Optional[str] = Field(default=None, max_length=100)
    cliente: Optional[str] = Field(default=None, max_length=200)
    condicion: Optional[str] = Field(default=None, max_length=200)
    observaciones_solicitante: Optional[str] = Field(default=None)
    placa_ficha: Optional[str] = Field(default=None, max_length=100)
    fecha_proximo_mantenimiento: Optional[date] = Field(default=None)

    # Gestión interna
    estado: str = Field(default=EstadoOC.nueva, max_length=30)
    # Referencia al user.id de intranet.db — sin FK constraint (DBs separadas)
    auxiliar_id: Optional[int] = Field(default=None)

    # Evidencia del solicitante (URL a OneDrive/SharePoint, enviada por PA)
    evidencia_url: Optional[str] = Field(default=None)

    # Campos de gestión de compras
    plataforma: Optional[str] = Field(default=None, max_length=100)
    numero_remision: Optional[str] = Field(default=None, max_length=100)
    observaciones_compras: Optional[str] = Field(default=None)
    fecha_estimada_entrega: Optional[date] = Field(default=None)
    fecha_confirmada_entrega: Optional[date] = Field(default=None)
    numero_factura: Optional[str] = Field(default=None, max_length=100)
    aval_compra: Optional[str] = Field(default=None, max_length=200)
    observacion_contabilidad: Optional[str] = Field(default=None)
    fecha_recibida_factura: Optional[date] = Field(default=None)

    # Fechas del proceso
    fecha_solicitud: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    fecha_asignacion: Optional[datetime] = Field(default=None)
    fecha_cotizacion: Optional[datetime] = Field(default=None)
    fecha_aprobacion: Optional[datetime] = Field(default=None)
    fecha_envio_oc: Optional[datetime] = Field(default=None)
    fecha_en_plataforma: Optional[datetime] = Field(default=None)
    fecha_recibido: Optional[datetime] = Field(default=None)

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class CotizacionProveedor(SQLModel, table=True):
    __tablename__ = "oc_cotizaciones"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    solicitud_id: uuid.UUID = Field(foreign_key="oc_solicitudes.id")

    # Datos del proveedor y cotización
    proveedor_nombre: str = Field(max_length=200)
    proveedor_nit: Optional[str] = Field(default=None, max_length=50)
    proveedor_email: Optional[str] = Field(default=None, max_length=200)
    numero_cotizacion_proveedor: Optional[str] = Field(default=None, max_length=100)
    valor_unitario: float
    valor_antes_iva: Optional[float] = Field(default=None)
    valor_iva: Optional[float] = Field(default=None)
    valor_total: float
    fecha_vigencia: Optional[date] = Field(default=None)
    forma_pago: Optional[str] = Field(default=None, max_length=200)
    plazo_entrega: Optional[str] = Field(default=None, max_length=200)
    observaciones: Optional[str] = Field(default=None)

    # Condiciones adicionales
    garantia: Optional[str] = Field(default=None, max_length=300)
    anticipo: Optional[str] = Field(default=None, max_length=200)
    pago_saldo: Optional[str] = Field(default=None, max_length=200)

    # PDF adjunto
    pdf_path: Optional[str] = Field(default=None)
    extraccion_automatica: bool = Field(default=False)

    # Aprobación
    aprobada: Optional[bool] = Field(default=None)
    valor_aprobado: Optional[float] = Field(default=None)
    # Referencia al user.id de intranet.db — sin FK constraint (DBs separadas)
    aprobado_por_id: Optional[int] = Field(default=None)
    observaciones_aprobacion: Optional[str] = Field(default=None)

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class OrdenCompra(SQLModel, table=True):
    __tablename__ = "oc_ordenes"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    solicitud_id: uuid.UUID = Field(foreign_key="oc_solicitudes.id")
    cotizacion_id: uuid.UUID = Field(foreign_key="oc_cotizaciones.id")

    numero_oc: str = Field(max_length=50)
    pdf_path: Optional[str] = Field(default=None)

    enviada_proveedor: bool = Field(default=False)
    enviada_coordinador: bool = Field(default=False)
    email_proveedor: Optional[str] = Field(default=None, max_length=200)

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class Proveedor(SQLModel, table=True):
    __tablename__ = "oc_proveedores"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    nombre: str = Field(max_length=200, sa_column_kwargs={"unique": True})
    email: Optional[str] = Field(default=None, max_length=200)
    telefono: Optional[str] = Field(default=None, max_length=50)
    nit: Optional[str] = Field(default=None, max_length=50)
    categoria: Optional[str] = Field(default=None, max_length=100)
    activo: bool = Field(default=True)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class OcConfig(SQLModel, table=True):
    """Configuración runtime de OC (SMTP, destinatarios). Sobreescribe .env."""
    __tablename__ = "oc_config"

    key: str = Field(primary_key=True, max_length=100)
    value: str = Field(default="")
